pre-docs/writing_yaml_plans.md

```diff
@@ -165,91 +142,214 @@ steps:
 
-Example:
+```yaml
+parameters:
+  users:
+    type: Array[String]
+
+steps:
+  - task: user::add
+    parameters:
+      name: $users.first
+  - task: echo
+    message: "hello ${users.join(',')}"
+```
+
+### Blocks
+
+Some Puppet functions take a block of code as an argument.
+
+For instance, you can filter an array of items based on the result of a block of code.
+
+```yaml
+parameters:
+  numbers:
+    type: Array[Integer]
+
+steps:
+  - task: sum
+    description: "add up the numbers > 5"
+    parameters:
+      indexes: $numbers.filter |$num| { $num > 5 }
+```
+
+## Connecting steps
+
+You can connect multiple steps by using the result of one step to compute the parameters for another step.
+
+### Using step results
+
+If a step has a `name` key, its result will be available to later steps in a variable with that name.
 
 ```yaml
 parameters:
   nodes:
     type: TargetSpec
-  message:
-    type: String
 
 steps:
-  - task: echo
+  - name: hostnames
+    command: hostname -f
     target: $nodes
+  - task: echo
     parameters:
-      message: $message
+      message: $hostnames.map |$hostname_result| { $hostname_result['stdout'] }.join(',')
 ```
 
-This plan will run the `echo` task on the targets specified in the `nodes` plan parameter, passing the `message` plan parameter to the task.
+This example uses the `map` function to get the value of `stdout` from each command result and then joins them into a single string separated by commas.
 
-You can also call Puppet functions from these expressions.
+### Eval step
 
-Example:
+You can use an `eval` step to evaluate an expression and save the result in a variable. This is useful to compute a variable to use multiple times later.
 
 ```yaml
 parameters:
-  message:
-    type: String
+  count:
+    type: Integer
 
 steps:
-  - task: echo
-    target: |
-      puppetdb_query('inventory[certname] { facts.osfamily = "Windows" }')
-    description: "Echo the message on the nodes matching a PuppetDB query"
-    parameters:
-      message: $message
+  - name: double_count
+    eval: $count * 2
 ```
 
+## Returning results
+
+You can return a result from a plan by setting the `return` key at the top level of the plan. When the plan finishes, the `return` key will be evaluated and returned as the result of the plan. If no `return` key is set, the plan will return `undef`.
+
+The `return` value can be a plain value or it can be a code expression.
+
+Example:
+```yaml
+steps:
+  - name: hostnames
+    command: hostname -f
+    target: $nodes
+
+return: $hostnames.map |$hostname_result| { $hostname_result['stdout'] }
+```
+
+## Advanced Puppet code expressions
+
+You can use a Puppet code expression as the value of any field of a step except the `name`. This allows for complex values to be computed.
+
+Immediately before running a step, any expressions embedded in the step will be evaluated. Each plan parameter and the values of every previous named step will be available in scope.
```

Comment:
I think this is a good point to stress the difference between loading and executing.

something like:
Bolt loads the yaml data structure. As it executes each step it will evaluate embedded expressions before running the step.